### PR TITLE
bringing sitemap and menu to CoreWebForm

### PR DESCRIPTION
**Pull request by birojnayak**
bringing sitemap and menu to CoreWebForm.
Initially it would bring basic required components for #261

### CODE REVIEW

File: src/WebForms/SiteMapProvider.cs [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+            {
+                // check the RawUrl without querystring
+                result = FindSiteMapNode(rawUrl.Substring(0, queryStringIndex));
+            }
+
+            if (result == null)
+            {
+                Page page = context.GetRequiredFeature<IHttpHandlerFeature>().Current as Page;
+
+                //TODO check with @taylor
```
Comment (birojnayak) [inline src/WebForms/SiteMapProvider.cs:176]: @twsouthwick take a look

File: src/WebForms/UI/Control.cs [hunk 2]
```diff
@@ -370,6 +371,50 @@ public virtual string ClientID
...
+            return null;
+        }
+
+        HttpContext context = Context;
+        if (context != null && context.Request.Browser != null)
+        {
+            AdapterInternal = null;
+
+//TODO create issue
+#if   GetAdapter_BrowserCapabilities
```
Comment (birojnayak) [inline src/WebForms/UI/Control.cs:392]: @twsouthwick should we handle here or in the adapter ?

File: src/WebForms/UI/PageAdapter.cs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+    internal string QueryString
+    {
+        get
+        {
+            string queryString = Page.ClientQueryString;
+
+          // https://github.com/dotnet/systemweb-adapters/issues/497
+
+     //     if (Page.Request.Browser.RequiresUniqueFilePathSuffix)
```
Comment (birojnayak) [inline src/WebForms/UI/PageAdapter.cs:65]: will create an internal issue

File: src/WebForms/SiteMapProvider.cs (hunk 1 above)
Comment (twsouthwick) [reply src/WebForms/SiteMapProvider.cs:176]: Can you add an extension method for `Page? GetCurrentPage(this HttpContext context)`?

File: src/WebForms/UI/Control.cs (hunk 2 above)
Comment (twsouthwick) [reply src/WebForms/UI/Control.cs:392]: create an issue here and mark this with PORT_BROWSERCAPABILITIES  - I like the `PORT_` prefix for easily identifiying what's needed

File: src/WebForms/UI/PageAdapter.cs (hunk 3 above)
Comment (twsouthwick) [reply src/WebForms/UI/PageAdapter.cs:65]: same as above

File: src/WebForms/Compilation/Polyfills.cs [hunk 4]
```diff
@@ -27,6 +30,75 @@ internal static object ParseExpression(string fullResourceKey)
...
+        return GetResourceObject(resourceProvider, resourceKey, culture,
+            objType, propName);
+    }
+
+    private static IResourceProvider GetGlobalResourceProvider(string classKey)
+    {
+        return null;
+
+#if PORT_GetGlobalResourceProvider
+        // throw new NotImplementedException();
```
Comment (twsouthwick) [inline src/WebForms/Compilation/Polyfills.cs:53]: for times like this, use HttpRuntimeHelper.GetLogger<> to get a logger to message things out

File: src/WebForms/Configuration/ConfigUtil.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
+//MIT license
+
+using System.Configuration;
+using System.Diagnostics;
+using System.Xml;
+
+namespace System.Web.Configuration
```
Comment (twsouthwick) [inline src/WebForms/Configuration/ConfigUtil.cs:7]: I'd prefer not to bring in any of the config. Is this just for SiteMapSection?

File: src/WebForms/Configuration/SiteMapSection.cs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        <providers>
+            <add name="AspNetXmlSiteMapProvider"
+                 description="SiteMap provider which reads in .sitemap XML files."
+                 type="System.Web.XmlSiteMapProvider, System.Web, Version=%ASSEMBLY_VERSION%, Culture=neutral, PublicKeyToken=%MICROSOFT_PUBLICKEY%"
+                 siteMapFile="web.sitemap" />
+        </providers>
+    </siteMap>
+*/
+
+public sealed class SiteMapSection : ConfigurationSection
```
Comment (twsouthwick) [inline src/WebForms/Configuration/SiteMapSection.cs:35]: make this internal, and I'd like to see a follow up that removes the config stuff and replaces it with an option

**Comment by twsouthwick** [comment]
@birojnayak are you still planning on pushing this through?

**Comment by birojnayak** [comment]
@twsouthwick yes.. give me sometime.

File: src/WebForms/Configuration/ProvidersHelper.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+using System.Collections.Specialized;
+using System.Configuration;
+using System.Configuration.Provider;
+using System.Reflection;
+
+namespace System.Web.Configuration
+{
+
+    public static class ProvidersHelper {
```
Comment (twsouthwick) [inline src/WebForms/Configuration/ProvidersHelper.cs:11]: I don't want any of the config stuff brought in. As an initial port, bringing it in as internal is fine but make sure to follow up with removing it

File: src/WebForms/Extensions.cs [hunk 8]
```diff
@@ -56,4 +57,6 @@ public static VirtualPath CurrentExecutionFilePathObject(this HttpRequest reques
     {
         return request.CurrentExecutionFilePath;
     }
+
+    public static Page GetCurrentPage(this HttpContext context) => context.GetRequiredFeature<IHttpHandlerFeature>().Current as Page;
```
Comment (twsouthwick) [inline src/WebForms/Extensions.cs:61]: you can use `context.CurrentHandler` now

File: test/Compiler.Dynamic.Tests/DynamicCompilationTests.cs [hunk 9]
```diff
@@ -159,7 +160,15 @@ public async Task CompiledPageRuns(string test, params string[] pages)
             File.WriteAllText(tempPath, result);
             _context.WriteLine($"Wrote result to {tempPath}");
 
-            Assert.AreEqual(expectedHtml.ReplaceLineEndings(), result.ReplaceLineEndings());
+            //TODO avoid this check, needed for Viewstate
```
Comment (twsouthwick) [inline test/Compiler.Dynamic.Tests/DynamicCompilationTests.cs:163]: can you create an issue to follow up here?

**Review by twsouthwick — CHANGES_REQUESTED** [review]
just a couple of visibility things then LGTM

**Comment by twsouthwick** [comment]
Please make sure the created issues aren't just "can you create an issue for this" - specify what needs to be done and link it to the tracking issue

**Comment by birojnayak** [comment]
> Please make sure the created issues aren't just "can you create an issue for this" - specify what needs to be done and link it to the tracking issue

read the title of the issue... :)

**Comment by twsouthwick** [comment]
Yes but the content of the issue should be standalone. Still has my request in there which does not make sense to have

**Comment by twsouthwick** [comment]
Also on such a large pr force pushes make it difficult to do incremental reviews - we'll squash at the end so just pushing fixes allow me to quickly verify things are updated

**Comment by birojnayak** [comment]
> Also on such a large pr force pushes make it difficult to do incremental reviews - we'll squash at the end so just pushing fixes allow me to quickly verify things are updated. I now have to go through the whole thing again

will keep this in mind @twsouthwick ..

**Pull request merged**